llvm/lib/Target/EVM/EVMCalculateModuleSize.cpp

```diff
@@ -0,0 +1,137 @@
+    // If it is not a simple branch, we are in a table somewhere.
+    if (!MI.isBranch() || MI.isIndirectBranch())
+      return false;
+
+    // If we are the operands of one of the branches, this is not a fall
+    // through. Note that targets with delay slots will usually bundle
+    // terminators with the delay slot instruction.
+    for (ConstMIBundleOperands OP(MI); OP.isValid(); ++OP) {
+      if (OP->isJTI())
+        return false;
+      if (OP->isMBB() && OP->getMBB() == MBB)
+        return false;
+    }
+  }
+
+  return true;
+}
+
+// Return the size of an instruction in bytes. For some pseudo instructions,
+// don't use the size from the instruction description, since during code
+// generation, some instructions will be relaxed to smaller instructions.
+static unsigned getInstSize(const MachineInstr &MI,
+                            const TargetInstrInfo *TII) {
+  // Skip debug instructions.
+  if (MI.isDebugInstr())
+    return 0;
+
+  unsigned Size = 0;
+  switch (MI.getOpcode()) {
+  default:
+    Size = MI.getDesc().getSize();
+    break;
+  case EVM::PseudoCALL:
+    // In case that function call has a return label, we will emit JUMPDEST,
+    // so take it into account.
+    if (MI.getNumExplicitOperands() > 1)
+      Size = TII->get(EVM::JUMPDEST_S).getSize();
+    LLVM_FALLTHROUGH;
+  case EVM::PseudoJUMPI:
+  case EVM::PseudoJUMP:
+    // Even though we are emitting PUSH4_S, most probably this will be relaxed
+    // to PUSH2_S when generating object code. In case we need larger
+    // instruction, bytecode limit is exceeded and this doesn't matter.
+    Size += TII->get(EVM::PUSH2_S).getSize() + TII->get(EVM::JUMP_S).getSize();
+    break;
+  case EVM::PUSH_LABEL:
+    // Even though we are emitting PUSH4_S, most probably this will be relaxed
+    // to PUSH2_S when generating object code. In case we need larger
+    // instruction, bytecode limit is exceeded and this doesn't matter.
+    Size = TII->get(EVM::PUSH2_S).getSize();
+    break;
+  }
+  return Size;
+}
+
+uint64_t llvm::EVM::calculateFunctionCodeSize(const MachineFunction &MF) {
+  uint64_t Size = 0;
+  const auto *TII = MF.getSubtarget<EVMSubtarget>().getInstrInfo();
+
+  // If the function has a PUSHDEPLOYADDRESS, it starts with a PUSH20.
+  if (const auto *MFI = MF.getInfo<EVMMachineFunctionInfo>();
+      MFI->getHasPushDeployAddress())
+    Size += TII->get(EVM::PUSH20).getSize();
+
+  for (const MachineBasicBlock &MBB : MF) {
+    // If the block is not only reachable by fallthrough, it starts with
+    // a JUMPDEST instruction.
+    if (!isBlockOnlyReachableByFallthrough(&MBB))
+      Size += TII->get(EVM::JUMPDEST_S).getSize();
+
+    Size += std::accumulate(MBB.begin(), MBB.end(), 0,
+                            [&TII](unsigned Sum, const MachineInstr &MI) {
+                              return Sum + getInstSize(MI, TII);
+                            });
+  }
+  return Size;
+}
+
+uint64_t llvm::EVM::calculateModuleCodeSize(Module &M,
```

Comment:
Dow we want a unit test for this?

Reply:
Sure, added: llvm/unittests/Target/EVM/BytecodeSize.cpp